src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs

```diff
@@ -0,0 +1,138 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    /// <summary>
+    /// Manages properties of analyzers (such as registered actions, supported diagnostics) for analyzer host's lifetime.
+    /// 
+    /// It ensures the following for the lifetime of analyzer host:
+    /// 1) <see cref="DiagnosticAnalyzer.Initialize(AnalysisContext)"/> is invoked only once per-analyzer.
+    /// 2) <see cref="DiagnosticAnalyzer.SupportedDiagnostics"/> is invoked only once per-analyzer.
+    /// 3) <see cref="CompilationStartAnalyzerAction"/> registered during Initialize are invoked only once per-analyzer per-compilation.
+    /// </summary>
+    internal class AnalyzerManager
+    {
+        public static readonly AnalyzerManager Default = new AnalyzerManager();
+
+        // This map stores the tasks to compute HostSessionStartAnalysisScope for session wide analyzer actions, i.e. AnalyzerActions registered by analyzer's Initialize method.
+        // These are run only once per every analyzer.
+        private ImmutableDictionary<DiagnosticAnalyzer, Task<HostSessionStartAnalysisScope>> _sessionScopeMap =
+            ImmutableDictionary<DiagnosticAnalyzer, Task<HostSessionStartAnalysisScope>>.Empty;
+
+        // This map stores the tasks to compute HostCompilationStartAnalysisScope for per-compilation analyzer actions, i.e. AnalyzerActions registered by analyzer's CompilationStartActions.
+        // Compilation start actions will get executed once per-each compilation as user might want to return different set of custom actions for each compilation.
+        private readonly ConditionalWeakTable<Compilation, ConcurrentDictionary<DiagnosticAnalyzer, Task<HostCompilationStartAnalysisScope>>> _compilationScopeMap =
+            new ConditionalWeakTable<Compilation, ConcurrentDictionary<DiagnosticAnalyzer, Task<HostCompilationStartAnalysisScope>>>();
+
+        /// <summary>
+        /// Cache descriptors for each diagnostic analyzer. We do this since <see cref="DiagnosticAnalyzer.SupportedDiagnostics"/> is
+        /// a property rather than metadata. We expect it to be cheap and immutable, but we can't force them to be so, we cache them
+        /// and ask only once.
+        /// </summary>
+        private readonly ConditionalWeakTable<DiagnosticAnalyzer, IReadOnlyList<DiagnosticDescriptor>> _descriptorCache =
+            new ConditionalWeakTable<DiagnosticAnalyzer, IReadOnlyList<DiagnosticDescriptor>>();
+
+        private Task<HostCompilationStartAnalysisScope> GetCompilationAnalysisScopeAsync(
```

Comment:
I feel this method a bit weird since it belong to analyzer manager. but only state it uses that belong to the analyzer manager is _compilationScopeMap. and it gets everything that belong to analyzer driver as argument.

feels like this class is not where this method is belong to. rather it belong to AnalyzerDriver and how data should flow is reversed.